# <div style = "font-size : 42px; color : #000000 ; font-family : 'Oregon'; text-align : center; background-color : #dba514; border-radius: 5px 5px;"><strong>Hotel Booking Cancellation Prediction</strong></div>

<div align="center">
  <img align = 'center' style = "border:5px solid #ffb037;" src="https://5.imimg.com/data5/EF/GO/MY-17287433/hotel-bookings-500x500.jpg" alt="Project Banner" width="1000"/>
</div>

## <a id = '0'></a>
<p style = "font-size : 35px; color : #f9858b ; font-family : 'Calibri'; text-align : center; background-color : #bdfff6; border-radius: 5px 5px;"><strong>Table of Content</strong></p> 

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

* [Table of Content](#0)
    * [Problem Statement](#0.1)
    * [About Notebook](#0.2)
    * [About Data](#0.3)
    * [Data Collection](#0.4)
    * [Data Description](#0.5)

* [Importing Libraries](#1)
    * [Load Dataset](#1.1)
    * [Look into Data](#1.2)
    * [Summary of Nummerical Features](#1.3)
    * [summary of Categorical Features](#1.4)
    * [Distribution of Categorical Features](#1.5)
    * [Distribution of Target Feature](#1.6)
 
* [Asking Key Questions](#2)
    * [1. Country wise Data distribution of guests who booked hotel successfully.](#2.1)
    * [2. How much do guests pay for a room per night?](#2.2)
    * [3. How does the price vary per night over the year?](#2.3)
    * [4. Which are the most busy months?](#2.4)
    * [5. How long do people stay at the hotels?](#2.5)

* [Ending Note](#3)
    * [Exploratory Data Analysis (EDA) Summary:](#3.1)


### <a id = '0.1'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Problem Statement</strong></div> 

<div style = "font-size : 16px; color : #ff9900; font-family : 'Comic Sans MS';">
    <strong>Problem:</strong> Client wants to reduce the cancellations of hotel bookings.<br>
    <strong>Solution:</strong> Using EDA techniques we have to find relation between features, hidden patterns and actionable insights from data.<br>
    Using machine learning algorithms we have to predict cancellations and implement targeted marketing strategies.

⬆️ [Tabel of Contents](#0)

### <a id = '0.2'></a>
<p style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>About Notebook</strong></p> 

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

This notebook conducts an `Exploratory Data Analysis (EDA)` on the hotel booking dataset, aiming to uncover key patterns, distributions, and relationships within the data. Through detailed visualizations and summary statistics, we aim to gain insights that will guide further data processing and model development. Key areas of exploration include:

- `Understanding Feature Distributions`: Analyzing numerical and categorical feature distributions to identify central tendencies, spread, and potential outliers.
- `Assessing Target Variable Patterns`: Examining the distribution of booking cancellations to understand customer behavior.
- `Identifying Correlations`: Investigating relationships between features to inform feature selection and potential engineering.
- `Segmentation Analysis`: Segmenting data by hotel type, country, or other relevant factors to reveal unique trends.
- `Asking Key Questions`: Formulating targeted questions to uncover meaningful patterns, such as which months have the highest cancellation rates, or how lead time correlates with booking behavior. This helps in identifying actionable insights that may not be immediately obvious from initial analysis.

This EDA provides a comprehensive overview of the dataset, guiding the next steps in data preprocessing and model building by highlighting essential insights and potential challenges in the data.

⬆️ [Tabel of Contents](#0)

### <a id = '0.3'></a>
<p style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>About Data</strong></p> 

<div style = "font-size : 16px; color : #ff9900; font-family : 'Comic Sans MS';">
    The data includes detailed information on hotel bookings, covering customer demographics, booking patterns, and reservation specifics.<br> 
    Key attributes include booking status, stay duration, guest count, booking channel, room assignment, and any special requests.<br> 
    It is suitable for analyzing booking trends, customer behaviors, and factors influencing cancellations and modifications.

⬆️ [Tabel of Contents](#0) 

### <a id = '0.4'></a>
<p style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Data Collection</strong></p> 

<div style = "font-size : 16px; color : #ff9900; font-family : 'Comic Sans MS';">
    The Hotel booking data is stored in mysql database we will fetch the data from the database<br>


⬆️ [Tabel of Contents](#0) 

### <div id = '0.5' style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Dataset Description</strong>

<div align='center' style="border-radius:10px; padding: 15px; font-size:15px;">

| __Index__ | __Variable__ | __Description__ |
|   :---    |     :---     |       :---      |
| 1 | __hotel__ | Type of hotel (Resort Hotel, City Hotel) |
| 2 | __is_canceled__ | Reservation cancellation status (0 = not canceled, 1 = canceled) |
| 3 | __lead_time__ | Number of days between booking and arrival |
| 4 | __arrival_date_year__ | Year of arrival |
| 5 | __arrival_date_month__ | Month of arrival |
| 6 | __arrival_date_week_number__ | Week number of the year for arrival |
| 7 | __arrival_date_day_of_month__ | Day of the month of arrival |
| 8 | __stays_in_weekend_nights__ | Number of weekend nights (Saturday and Sunday) the guest stayed or booked |
| 9 | __stays_in_week_nights__ | Number of week nights the guest stayed or booked |
| 10 | __adults__ | Number of adults |
| 11 | __children__ | Number of children |
| 12 | __babies__ | Number of babies |
| 13 | __meal__ | Type of meal booked (BB, FB, HB, SC, Undefined) |
| 14 | __country__ | Country of origin of the guest |
| 15 | __market_segment__ | Market segment designation |
| 16 | __distribution_channel__ | Booking distribution channel |
| 17 | __is_repeated_guest__ | If the guest is a repeat customer (0 = not repeated, 1 = repeated) |
| 18 | __previous_cancellations__ | Number of previous bookings that were canceled by the customer |
| 19 | __previous_bookings_not_canceled__ | Number of previous bookings that were not canceled by the customer |
| 20 | __reserved_room_type__ | Type of reserved room |
| 21 | __assigned_room_type__ | Type of assigned room |
| 22 | __booking_changes__ | Number of changes made to the booking |
| 23 | __deposit_type__ | Type of deposit made (No Deposit, Refundable, Non Refund) |
| 24 | __agent__ | ID of the travel agent responsible for the booking |
| 25 | __company__ | ID of the company responsible for the booking |
| 26 | __days_in_waiting_list__ | Number of days the booking was in the waiting list |
| 27 | __customer_type__ | Type of customer (Transient, Contract, Transient-Party, Group) |
| 28 | __adr__ | Average Daily Rate |
| 29 | __required_car_parking_spaces__ | Number of car parking spaces required |
| 30 | __total_of_special_requests__ | Number of special requests made |
| 31 | __reservation_status__ | Last reservation status (Check-Out, Canceled, No-Show) |
| 32 | __reservation_status_date__ | Date of the last reservation status |
| 33 | __name__ | Guest's name |
| 34 | __email__ | Guest's email address |
| 35 | __phone-number__ | Guest's phone number |
| 36 | __credit_card__ | Last four digits of the guest's credit card |
</div>

⬆️ [Tabel of Contents](#0)

### <a id = '1'></a>
<div style = "font-size : 40px; color : #f9858b ; font-family : 'Calibri'; text-align : center; background-color : #bdfff6; border-radius: 5px 5px;"><strong>Importing Libraries</strong></div> 

In [1]:
import warnings
warnings.filterwarnings("ignore")

# database connection
from sqlalchemy import create_engine

# data manipulation
import pandas as pd
import numpy as np

# visualization
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
from folium.plugins import HeatMap

# style
plt.style.use('fivethirtyeight')
%matplotlib inline
pd.set_option('display.max_columns', 40)

### <a id = '1.1'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Load Dataset</strong></div> 

⬆️ [Tabel of Contents](#0)

In [2]:
engine = create_engine("mysql+mysqlconnector://projects:AIMLprojects1@127.0.0.1:3306/projects_db")
conn = engine.connect()

In [3]:
query = "SELECT * FROM hotel_booking"

In [4]:
df = pd.read_sql(query, engine)
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent company  \
0                  C                3   No Deposit                  
1                  C                4   No Deposit                  
2                  C                0   No Deposit                  
3                  A                0   No Deposit  304.0           
4                  A                0   No Deposit  240.0           

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  \
0                          0          Check-Out              2015-07-01   
1                          0          Check-Out              2015-07-01   
2                          0          Check-Out              2015-07-02   
3                          0          Check-Out              2015-07-02   
4                          1          Check-Out              2015-07-03   

             name                        email  phone-number       credit_card  
0   Ernest Barnes  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1    Andrea Baker       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2  Rebecca Parker   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3    Laura Murray            Laura_M@gmail.com  364-656-8427  ************5677  
4     Linda Hines           LHines@verizon.com  713-226-5

### <a id = '1.2'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Look into Data</strong></div> 

⬆️ [Tabel of Contents](#0)

In [5]:
df.shape

(119386, 36)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119386 entries, 0 to 119385
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119386 non-null  object 
 1   is_canceled                     119386 non-null  int64  
 2   lead_time                       119386 non-null  int64  
 3   arrival_date_year               119386 non-null  int64  
 4   arrival_date_month              119386 non-null  object 
 5   arrival_date_week_number        119386 non-null  int64  
 6   arrival_date_day_of_month       119386 non-null  int64  
 7   stays_in_weekend_nights         119386 non-null  int64  
 8   stays_in_week_nights            119386 non-null  int64  
 9   adults                          119386 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119386 non-null  int64  
 12  meal            

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insights:</font></h4>

* __Number of Entries__: The dataset consists of __119,390 entries__.

* __Columns__: The dataset contains __36 columns__, which represent various attributes related to hotel bookings.

* __Data Types__:
    - A majority of the columns, 16 to be precise, are of the __object__ data type (often representing strings or categorical data).
    - 16 columns are of the __int64__ data type, representing integer values.
    - 4 columns are of the __float64__ data type, which typically denotes decimal values.

* __Missing Values__: 
    - The column `children` has 4 missing values.
    - The column `country` has 488 missing values.
    - The column `agent` has 16,340 missing values.
    - The column `company` has a significant number of missing values, totaling 112,593.

These missing values might need addressing, depending on the context of the analysis or the model we intend to build.

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Note:</font></h4>
    
Based on the data types and the feature explanations provided earlier, we identified that __20 columns__ (`hotel`, `is_canceled`, `arrival_date_year`, `arrival_date_month`, `meal`, `country`, `market_segment`, `distribution_channel`, `is_repeated_guest`, `reserved_room_type`, `assigned_room_type`, `deposit_type`, `agent`, `company`, `customer_type`, `reservation_status`, `name`, `email`, `phone-number` and `credit_card`) are __categorical__ in terms of their semantics. These features must have string (__object__) data type to ensure proper analysis and interpretation in subsequent steps:

⬆️ [Tabel of Contents](#0)

In [7]:
# Based on the given dataset description, list the columns that are categorical in terms of their semantics
categorical_columns_semantics = [
    'hotel',
    'is_canceled',
    'arrival_date_year',
    'arrival_date_month',
    'meal',
    'country',
    'market_segment',
    'distribution_channel',
    'is_repeated_guest',
    'reserved_room_type',
    'assigned_room_type',
    'deposit_type',
    'agent',
    'company',
    'customer_type',
    'reservation_status',
    'name',
    'email',
    'phone-number',
    'credit_card'
]

# Convert these columns to string (object) data type
for column in categorical_columns_semantics:
    if column in df.columns:
        df[column] = df[column].astype(str)

# Verify the changes in data types
df.dtypes

hotel                              object
is_canceled                        object
lead_time                           int64
arrival_date_year                  object
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                  object
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

### <a id = '1.3'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Summary of Numrical Features</strong></div> 

In [8]:
df.describe().T

count        mean         std   min    25%  \
lead_time                       119386.0  104.014801  106.863286  0.00  18.00   
arrival_date_week_number        119386.0   27.165003   13.605334  1.00  16.00   
arrival_date_day_of_month       119386.0   15.798553    8.780783  1.00   8.00   
stays_in_weekend_nights         119386.0    0.927605    0.998618  0.00   0.00   
stays_in_week_nights            119386.0    2.500310    1.908289  0.00   1.00   
adults                          119386.0    1.856390    0.579261  0.00   2.00   
children                        119386.0    0.103890    0.398561  0.00   0.00   
babies                          119386.0    0.007949    0.097438  0.00   0.00   
previous_cancellations          119386.0    0.087121    0.844350  0.00   0.00   
previous_bookings_not_canceled  119386.0    0.137102    1.497462  0.00   0.00   
booking_changes                 119386.0    0.221131    0.652315  0.00   0.00   
days_in_waiting_list            119386.0    2.321227   17.595011  0.00   0.00   
adr                             119386.0  101.833541   50.534664 -6.38  69.29   
required_car_parking_spaces     119386.0    0.062520    0.245295  0.00   0.00   
total_of_special_requests       119386.0    0.571340    0.792798  0.00   0.00   

                                  50%    75%     max  
lead_time                       69.00  160.0   737.0  
arrival_date_week_number        28.00   38.0    53.0  
arrival_date_day_of_month       16.00   23.0    31.0  
stays_in_weekend_nights          1.00    2.0    19.0  
stays_in_week_nights             2.00    3.0    50.0  
adults                           2.00    2.0    55.0  
children                         0.00    0.0    10.0  
babies                           0.00    0.0    10.0  
previous_cancellations           0.00    0.0    26.0  
previous_bookings_not_canceled   0.00    0.0    72.0  
booking_changes                  0.00    0.0    21.0  
days_in_waiting_list             0.00    0.0   391.0  
adr                             94.59  126.0  5400.0  
required_car_parking_spaces      0.00    0.0     8.0  
total_of_special_requests        0.00    1.0     5.0

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Numerical Features:</font></h4>

* __`lead_time`__: The average time between booking and arrival is approximately 104 days, ranging from 0 to 737 days.
    

* __`arrival_date_week_number`__: The average week number of the year for arrival is approximately 27.17, ranging from 1 to 53.

    
* __`arrival_date_day_of_month`__: The average day of the month of arrival is approximately 15.80, ranging from 1 to 31.
    
    
* __`stays_in_weekend_nights`__: On average, guests stay for around 0.93 weekend nights, with some guests staying for as long as 19 weekend nights.
    
    
* __`stays_in_week_nights`__: On average, guests stay for about 2.5 week nights, with stays ranging from 0 to 50 week nights.
    
    
* __`adults`__: The average number of adults per booking is approximately 1.86, with a range from 0 to 55 adults.
    
    
* __`children`__: On average, there are about 0.1 children per booking, with a maximum of 10 children in some bookings.
    
    
* __`babies`__: The average number of babies per booking is very low, approximately 0.008, with a maximum of 10 babies in some bookings.
    
    
* __`previous_cancellations`__: On average, guests have canceled about 0.09 times in the past, with some guests having as many as 26 previous cancellations.
    
    
* __`previous_bookings_not_canceled`__: On average, guests have not canceled about 0.14 bookings in the past, with some guests having as many as 72 previous bookings that they didn't cancel.
    
    
* __`booking_changes`__: Bookings are changed on average about 0.22 times, with some bookings being changed up to 21 times.
    
    
* __`days_in_waiting_list`__: The average time a booking spends on the waiting list is approximately 2.32 days, with some bookings waiting for as long as 391 days.
    
    
* __`adr`__: The Average Daily Rate is about 101.83, with rates ranging from -6.38 (which might indicate errors or special cases) to 5400.
    
    
* __`required_car_parking_spaces`__: On average, guests require about 0.06 parking spaces, with some bookings requiring up to 8 spaces.
    
    
* __`total_of_special_requests`__: On average, guests make about 0.57 special requests, with some making up to 5 requests.

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Noisy Data:</font></h4>

Based on the descriptions and the inferences provided earlier, the following features seem to contain noisy data:

1. __`adr`__: The Average Daily Rate (ADR) represents the average amount a guest pays for a room per day. In this dataset, ADR has a minimum value of -6.38. A negative rate doesn't make sense, indicating potential errors or special cases.

    
2. __`adults`__: The minimum value is 0, which implies bookings without any adults. This could be a data entry error unless there are legitimate cases where only children or babies book a room.

    
3. __`children`__ and __`babies`__: Both have a maximum value of 10, which seems unusually high for a single booking. This could be a potential outlier or data entry error, but it would require more context or domain knowledge to say definitively.

Further investigation, possibly involving domain expertise or additional context, would be necessary to determine the true nature of these data points.

⬆️ [Tabel of Contents](#0)

### <a id = '1.4'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Summary of categorical features</strong></div> 

In [9]:
# Summary of categorical features
df.describe(include='object').T

count  unique                  top    freq
hotel                    119386       2           City Hotel   79326
is_canceled              119386       2                    0   75166
arrival_date_year        119386       3                 2016   56707
arrival_date_month       119386      12               August   13873
meal                     119386       5                   BB   92306
country                  119386     178                  PRT   48586
market_segment           119386       7            Online TA   56476
distribution_channel     119386       5                TA/TO   97870
is_repeated_guest        119386       2                    0  115576
reserved_room_type       119386      10                    A   85994
assigned_room_type       119386      12                    A   74053
deposit_type             119386       3           No Deposit  104637
agent                    119386     334                  9.0   31960
company                  119386     353                       112589
customer_type            119386       4            Transient   89613
reservation_status       119386       3            Check-Out   75166
reservation_status_date  119386     926           2015-10-21    1461
name                     119386   81501      Michael Johnson      48
email                    119386  115885  Michael.C@gmail.com       6
phone-number             119386  119386         669-792-1661       1
credit_card              119386    9000     ************4923      28

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">
    
    
<h4 align="left"><font color=bdfff6>Categorical Features:</font></h4>

* __`hotel`__: There are two types of hotels, with "City Hotel" being the most frequent category, occurring 79,330 times out of 119,390 entries.

    
* __`is_canceled`__: Two unique values (0 for not canceled and 1 for canceled) are present. The most common value is "0" (not canceled), occurring 75,166 times.

    
* __`arrival_date_year`__: There are three unique years present, with 2016 being the most frequent year of arrival, observed 56,707 times.

    
* __`arrival_date_month`__: There are 12 unique months. August is the most frequent month of arrival, observed 13,877 times.

    
* __`meal`__: Five unique types of meals are booked. The most common type is "BB", observed 92,310 times.

    
* __`country`__: 178 unique countries of origin are present. The most frequent country is "PRT" (Portugal), observed 48,590 times.

    
* __`market_segment`__: Eight unique market segments are present. The most common segment is "Online TA", observed 56,477 times.

    
* __`distribution_channel`__: Five unique booking distribution channels are present, with "TA/TO" being the most frequent, observed 97,870 times.

    
* __`is_repeated_guest`__: Two unique values (0 for not repeated and 1 for repeated) are present. The most frequent value is "0" (not repeated), observed in the majority of entries.

    
* __`reserved_room_type`__ and __`assigned_room_type`__: Various room types are present, with some types being more frequent than others.

    
* __`deposit_type`__: Three unique types of deposits are made. The most common type is "No Deposit", observed 104,641 times.


* __`agent`__: There are 334 unique agents, with '9.0' being the most frequent, occurring 31,961 times.
    
    
* __`company`__: There are 353 unique companies, but '__nan__' (missing values) is the most frequent, occurring 112,593 times. This indicates that the 'company' variable has a high percentage of missing values.
    
    
* __`customer_type`__: Four unique customer types are present. The most frequent type is "Transient", observed 89,613 times.

    
* __`reservation_status`__: Three unique reservation statuses are present. The most common status is "Check-Out", observed 75,166 times.


* __`reservation_status_date`__: There are 926 unique values, with '2015-10-21' being the most frequent, occurring 1,461 times. 


* __`name`__: There are 81,503 unique names, with 'Michael Johnson' being the most frequent, occurring 48 times.
    

* __`email`__: There are 115,889 unique email addresses, with 'Michael.C@gmail.com' being the most frequent, occurring 6 times.
    
    
* __`phone-number`__: There are 119,390 unique phone numbers, indicating that almost every guest has a unique phone number.
    
    
* __`credit_card`__: There are 9,000 unique credit card numbers, with '************4923' being the most frequent, occurring 28 times.

⬆️ [Tabel of Contents](#0)

### <a id = '1.5'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Distribution of Categorical Features</strong></div> 

In [10]:
# define numerical & categorical columns
numeric_features = [_ for _ in df.columns if df[_].dtype != 'O']
categorical_features = [_ for _ in df.columns if df[_].dtype == 'O']

# print columns
print(f'We have {len(numeric_features)} numerical features : {numeric_features}')
print(f'\nWe have {len(categorical_features)} categorical features : {categorical_features}')

We have 15 numerical features : ['lead_time', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']

We have 21 categorical features : ['hotel', 'is_canceled', 'arrival_date_year', 'arrival_date_month', 'meal', 'country', 'market_segment', 'distribution_channel', 'is_repeated_guest', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'agent', 'company', 'customer_type', 'reservation_status', 'reservation_status_date', 'name', 'email', 'phone-number', 'credit_card']


In [11]:
# proportion of count data on categorical columns
for col in categorical_features:
    print(df[col].value_counts(normalize=True) * 100)
    print('- ' * 50)

hotel
City Hotel      66.444977
Resort Hotel    33.555023
Name: proportion, dtype: float64
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
is_canceled
0    62.960481
1    37.039519
Name: proportion, dtype: float64
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
arrival_date_year
2016    47.498869
2017    34.080210
2015    18.420920
Name: proportion, dtype: float64
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
arrival_date_month
August       11.620290
July         10.605096
May           9.876367
October       9.347830
April         9.288359
June          9.162716
September     8.801702
March         8.203642
February      6.757911
November      5.690785
December      5.679058
January       4.966244
Name: proportion, dtype: float64
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">
<h4 align="left"><font color="#bdfff6">Insights:</font></h4>

* __Hotel Types__: 
    - __`City Hotel`__ bookings dominate at about 66%, while __`Resort Hotel`__ bookings make up the remaining 34%.
    - This suggests more guests prefer city accommodations, likely due to accessibility or travel purposes.

* __Arrival Months__: 
    - The busiest months are __`August`__ (11.6%) and __`July`__ (10.6%), followed closely by __`May`__ and __`October`__.
    - This indicates a peak booking season in summer and a secondary rise in October, possibly due to tourism trends.

* __Meal Preferences__: 
    - __Breakfast-only (`BB`)__ is the most popular meal plan at 77%, followed by __Half-Board (`HB`)__ at 12%.
    - This indicates a preference for light, breakfast-focused dining over more inclusive plans.

* __Country of Origin__: 
    - Most guests are from __`Portugal` (40.7%)__, followed by __`the UK` (10.2%)__, __`France` (8.7%)__, __`Spain` (7.2%)__, and __`Germany` (6.1%)__.
    - This suggests a predominantly local (Portuguese) clientele, with a significant proportion from nearby European countries.

* __Market Segment__: 
    - The largest segment is __`Online Travel Agencies` (47%)__, followed by __`Offline TA/TO` (20%)__ and __`Groups` (16.6%)__.
    - This trend reflects the importance of online booking channels.

* __Distribution Channels__: 
    - The __Travel Agent/Operator (`TA/TO`)__ channel accounts for 82% of bookings, followed by __`Direct`__ bookings (12%) and __`Corporate`__ channels (5.6%).
    - This suggests reliance on travel agencies and operators.

* __Room Types__: 
    - Reserved room type __`A`__ is the most popular at 72%, but assigned room type __`A`__ only makes up 62%, indicating some reassignment.
    - Room types __`D`__ and __`E`__ are also common.

* __Deposit Type__: 
    - __`No Deposit`__ is chosen by 87.6% of guests, while __`Non-Refund`__ options account for 12%.
    - This shows a strong preference for flexibility.

* __Customer Type__: 
    - __`Transient`__ customers (75%) are the most common, likely reflecting short-term travelers.
    - __`Transient-Party`__ (21%) and __`Contract`__ (3.4%) types follow, with very few __`Group`__ bookings (0.5%).

* __Reservation Status__: 
    - __`Check-Out`__ accounts for 63% of reservations, __`Canceled`__ reservations make up 36%, and __`No-Show`__ is around 1%.
    - This indicates a significant number of cancellations.

* __Reservation Dates__: 
    - Some specific dates have a higher booking proportion, such as __`2015-10-21`__ and __`2015-07-06`__.
    - These may reflect high-demand days or holidays.

* __Unique Identifiers__: 
    - __`Name`__, __`email`__, __`phone number`__, and __`credit card`__ details each appear to be highly unique with very few repeating values, as expected for identifiers.

These insights can inform strategies around pricing, marketing (especially in peak months), channel prioritization, and customer engagement. 
For instance, focusing on online channels and understanding preferences for city hotels, breakfast-only options, and flexible booking policies could help optimize service offerings.

⬆️ [Tabel of Contents](#0)


### <a id = '1.6'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Distribution of Target Feature</strong></div> 

In [12]:
# Calculate percentages
percentage = df['is_canceled'].value_counts(normalize=True) * 100
labels = ["Not Cancelled", "Cancelled"]

# Create the pie chart using Plotly Express
fig = px.pie(
    values=percentage, 
    names=labels, 
    title="Cancellation Status",
    color_discrete_sequence=['#1188ff', '#e63a2a'],  # Set custom colors
    hole=0,  # Use 0 for a regular pie chart, >0 for a donut chart
    template="plotly_dark"  # Apply dark theme
)

fig.data[0].pull = [0, 0.1]  # Explode only the "Cancelled" slice

# Show the chart
fig.show()


<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insight: Cancellation Status </font></h4>

- **Majority of bookings are not canceled (`63%`)**: The data shows that the majority of bookings are completed as planned, which reflects a generally high level of guest commitment and hotel booking stability.
  
- **Substantial portion of bookings are canceled (`37%`)**: While a significant percentage of bookings are canceled, it still remains a substantial portion. This suggests that understanding the reasons behind these cancellations could be key to reducing them.

⬆️ [Tabel of Contents](#0)

## <a id = '2'></a>
<div style = "font-size : 35px; color : #f9858b ; font-family : 'Calibri'; text-align : center; background-color : #bdfff6; border-radius: 5px 5px;"><strong>Asking Key Questions</strong></div> 

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

Formulating targeted questions to uncover meaningful patterns, such as from which country most guests are from, or how much do guests pay for a night. This helps in identifying actionable insights that may not be immediately obvious from initial analysis.

⬆️ [Tabel of Contents](#0)

### <a id = '2.1'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>1. Country wise Data distribution of guests who booked hotel successfully.</strong></div> 

In [13]:
country_wise_data = df[df.is_canceled == '0']['country'].value_counts().reset_index()
country_wise_data.columns = ['Country', 'No of Guests']
country_wise_data.head()

Country  No of Guests
0     PRT         21071
1     GBR          9676
2     FRA          8481
3     ESP          6391
4     DEU          6069

In [14]:
fig = px.choropleth(
    country_wise_data,
    locations = 'Country',
    color = 'No of Guests',
    hover_name = 'Country',
    title = 'Country wise Guests',
    color_continuous_scale = 'Viridis',
    range_color = [0, max(country_wise_data['No of Guests'])],
    locationmode = 'ISO-3',
)

fig.show()

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insight: Country-wise Guest Distribution </font></h4>
    
The Map chart indicates that the majority of hotel guests are from Portugal (__`PRT`__), with __21,071__ guests, which suggests a strong domestic customer base.<br> 
The United Kingdom (__`GBR`__) and France (__`FRA`__) follow, with __9,676__ and __8,481__ guests, respectively, highlighting a significant international presence from nearby European countries.<br> 
Spain (__`ESP`__) and Germany (__`DEU`__) also contribute a notable number of guests, with __6,391__ and __6,069__ visitors, respectively.<br> 
This distribution reflects a dominant local and regional appeal, likely influenced by proximity, tourism, and cultural ties within Europe.

⬆️ [Tabel of Contents](#0)

### <a id = '2.2'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>2. How much do guests pay for a room per night?</strong></div> 

In [15]:
data = df[df['is_canceled'] == '0']

fig = px.box(
    data,
    x = 'reserved_room_type',
    y = 'adr',
    color = 'hotel',
    template = 'plotly_dark',
    labels = {'reserved_room_type': 'Reserved Room Type', 'adr' : 'Average Daily Rate'},
    title = 'Average pay per night'
)

fig.show()

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insights: Average pay per night</font></h4>

1. **Price Variability Across Room Types**:
   - Room types `C`, `A`, and `D` have a wider range of prices, with notable variability in ADR for both Resort and City hotels.
   - Room types `H` and `L` have higher median ADR values but with fewer data points, suggesting they may be premium or less frequently booked room types.

2. **Hotel Type Differences**:
   - In general, `Resort Hotel` rooms show higher variability in pricing compared to `City Hotel` rooms. This could indicate more diverse pricing based on factors like seasonality or room amenities.
   - For certain room types (like `C`), Resort Hotel has a significantly higher ADR spread, meaning it may cater to a range of budgets or demand fluctuates more for these rooms.

3. **Outliers**:
   - There are several outliers, particularly in room type `C` for both hotels, and room type `A` in the Resort Hotel. These outliers could represent peak-season rates or special circumstances affecting pricing.

4. **Median Comparison**:
   - For most room types (A, D, E), City Hotel’s median ADR is lower than Resort Hotel’s, implying Resort Hotels might be positioned as a more premium option.

In summary, the plot suggests that **Resort Hotels generally have higher pricing variability and higher median ADRs** for most room types. This could indicate that Resort Hotels are targeting a broader market segment or have pricing strategies based on seasonal demand.

⬆️ [Tabel of Contents](#0)

### <a id = '2.3'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>3. How does the price vary per night over the year?</strong></div> 

In [16]:
resort = df[(df['hotel'] == 'Resort Hotel') & (df['is_canceled'] == '0')]
city = df[(df['hotel'] == 'City Hotel') & (df['is_canceled'] == '0')]

In [17]:
resort_hotel = resort.groupby(['arrival_date_month'])['adr'].mean().reset_index()
resort_hotel

arrival_date_month         adr
0               April   75.867816
1              August  181.205892
2            December   68.322236
3            February   54.147478
4             January   48.708919
5                July  150.122528
6                June  107.921869
7               March   57.012487
8                 May   76.657558
9            November   48.681640
10            October   61.727505
11          September   96.416860

In [18]:
city_hotel = city.groupby(['arrival_date_month'])['adr'].mean().reset_index()
city_hotel

arrival_date_month         adr
0               April  111.856824
1              August  118.412083
2            December   87.856764
3            February   86.183025
4             January   82.160634
5                July  115.563810
6                June  117.702075
7               March   90.170722
8                 May  120.445842
9            November   86.500456
10            October  101.745956
11          September  112.598452

In [19]:
hotel = resort_hotel.merge(city_hotel, on = 'arrival_date_month')
hotel.columns = ['month', 'resort_hotel_price', 'city_hotel_price']
hotel

month  resort_hotel_price  city_hotel_price
0       April           75.867816        111.856824
1      August          181.205892        118.412083
2    December           68.322236         87.856764
3    February           54.147478         86.183025
4     January           48.708919         82.160634
5        July          150.122528        115.563810
6        June          107.921869        117.702075
7       March           57.012487         90.170722
8         May           76.657558        120.445842
9    November           48.681640         86.500456
10    October           61.727505        101.745956
11  September           96.416860        112.598452

<div style = "font-size : 16px; color : #ff9900; font-family : 'Comic Sans MS';">
    month column is not in order, if we visualize it we will not going to get good insights.<br>
    First we have to re-arrange month column by using 'sort_dataframeby_monthorweek' module.
</div> 

In [20]:
import sort_dataframeby_monthorweek as sd
def sort_month(df, column):
    return sd.Sort_Dataframeby_Month(df, column)

In [21]:
hotel_prices = sort_month(hotel, 'month')
hotel_prices

month  resort_hotel_price  city_hotel_price
0     January           48.708919         82.160634
1    February           54.147478         86.183025
2       March           57.012487         90.170722
3       April           75.867816        111.856824
4         May           76.657558        120.445842
5        June          107.921869        117.702075
6        July          150.122528        115.563810
7      August          181.205892        118.412083
8   September           96.416860        112.598452
9     October           61.727505        101.745956
10   November           48.681640         86.500456
11   December           68.322236         87.856764

In [22]:
fig = px.line(
    hotel_prices,
    x = 'month',
    y = ['resort_hotel_price', 'city_hotel_price'],
    title = 'Room price per night over the Months',
    template = 'plotly_dark'
)

fig.show()

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insights: Room price per night over the months</font></h4>

1. **Seasonal Price Trends**:
   - Both `Resort` and `City` Hotels show a clear **seasonal trend**, with prices peaking during the summer months (June to August).
   - `August` has the highest average price for both types of hotels, with Resort Hotels reaching `181.21` and City Hotels `118.41`. This is likely due to increased demand during the vacation season.

2. **Price Differences Between Hotel Types**:
   - `Resort Hotels` generally have a higher price than `City Hotels` in most months, especially from `June` to `September`. This could indicate that Resort Hotels experience higher demand or cater to more premium offerings during these months.
   - In the off-season months, such as `January`, `February`, and `November`, the price difference between Resort and City Hotels is much smaller, suggesting a more balanced demand during these months.

3. **Low-Season Rates**:
   - `January` and `November` have the lowest prices for both Resort and City Hotels, which aligns with typical low-season trends when tourism demand is generally lower.
   - `Resort Hotel` prices are consistently lower than City Hotels in January and November, indicating that resorts may have more flexibility in lowering prices to attract guests during off-peak times.

4. **City Hotels’ Price Stability**:
   - The prices for City Hotels are relatively stable throughout the year, with a moderate increase during peak seasons. This stability suggests that City Hotels might cater more to business travelers, who maintain demand throughout the year with less sensitivity to seasonality.

5. **Resort Hotels’ Price Volatility**:
   - Resort Hotels exhibit greater price fluctuations compared to City Hotels, with a significant increase in summer months and sharp declines in winter. This volatility reflects their stronger dependence on seasonal tourists, especially those seeking leisure travel.

**Summary**:
Resort Hotels experience higher seasonality with prices peaking in the summer months, driven by leisure travelers. In contrast, City Hotels show more stable pricing year-round, likely due to business travelers. This insight could be useful for pricing strategies, with Resort Hotels potentially benefiting from dynamic pricing models to capitalize on seasonal demand while City Hotels might focus on consistent pricing to appeal to business clientele.

⬆️ [Tabel of Contents](#0)

### <a id = '2.4'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>4. What are the most busy months?</strong></div> 

In [23]:
city_guest = city['arrival_date_month'].value_counts().reset_index()
city_guest.columns = ['month', 'No of Guests']
city_guest

month  No of Guests
0      August          5381
1        July          4782
2         May          4579
3        June          4366
4     October          4337
5   September          4290
6       March          4072
7       April          4015
8    February          3064
9    November          2696
10   December          2392
11    January          2254

In [24]:
resort_guest = resort['arrival_date_month'].value_counts().reset_index()
resort_guest.columns = ['month', 'No of Guests']
resort_guest

month  No of Guests
0      August          3257
1        July          3137
2     October          2577
3       March          2573
4       April          2550
5         May          2535
6    February          2308
7   September          2102
8        June          2038
9    December          2017
10   November          1976
11    January          1868

In [25]:
guest = resort_guest.merge(city_guest, on = 'month')
guest.columns = ['month', 'resort_hotel_guest', 'city_hotel_guest']
guest

month  resort_hotel_guest  city_hotel_guest
0      August                3257              5381
1        July                3137              4782
2     October                2577              4337
3       March                2573              4072
4       April                2550              4015
5         May                2535              4579
6    February                2308              3064
7   September                2102              4290
8        June                2038              4366
9    December                2017              2392
10   November                1976              2696
11    January                1868              2254

In [26]:
guest = sort_month(guest, 'month')
guest

month  resort_hotel_guest  city_hotel_guest
0     January                1868              2254
1    February                2308              3064
2       March                2573              4072
3       April                2550              4015
4         May                2535              4579
5        June                2038              4366
6        July                3137              4782
7      August                3257              5381
8   September                2102              4290
9     October                2577              4337
10   November                1976              2696
11   December                2017              2392

In [27]:
fig = px.line(
    guest,
    x = 'month',
    y = ['resort_hotel_guest', 'city_hotel_guest'],
    title = 'Guest distribution over the month',
    template = 'plotly_dark'
)

fig.show()

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insights: Guest distribution over the month</font></h4>

1. **Busiest Months**:
   - `August` is the busiest month for both hotel types, with `3,257` guests at Resort Hotels and `5,367` guests at City Hotels. This aligns with peak vacation season, where tourism is typically high.
   - `July` and `August` are consistently high for both hotels, confirming summer as the peak period.

2. **Seasonal Patterns**:
   - For both Resort and City Hotels, guest numbers peak during the summer months from `June` to `August`. 
   - After August, guest counts start to decline in `September`, but City Hotels maintain relatively high guest numbers until `October`, possibly due to business travel as summer vacations end.

3. **City vs. Resort Hotel Guest Counts**:
   - `City Hotels` consistently attract more guests than Resort Hotels in every month, indicating their broader appeal or higher capacity, possibly serving a mix of business and leisure travelers.
   - Resort Hotels show a noticeable decline in guests during the winter months, particularly in `January` and `November`, likely due to lower leisure travel demand.

4. **Least Busy Months**:
   - `January` and `November` are the least busy months for both hotel types. City Hotels see around `2,249` (`January`) and `2,676` (`November`) guests, while Resort Hotels drop to `1,866` (`January`) and `1,975` (`November`).

**Summary**:
`City Hotels` experience relatively steady demand throughout the year, with a clear increase in summer months. `Resort Hotels`, however, are much more seasonal, with high guest counts in summer and significantly lower numbers in winter. For both hotels, the **summer months (June to August)** are the busiest, especially **July and August**, due to increased vacation travel.

⬆️ [Tabel of Contents](#0)

### <a id = '2.5'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>5. How long do people stay in hotels?</strong></div> 

In [28]:
filter = df['is_canceled'] == '0'
data = df[filter]
data.head()

hotel is_canceled  lead_time arrival_date_year arrival_date_month  \
0  Resort Hotel           0        342              2015               July   
1  Resort Hotel           0        737              2015               July   
2  Resort Hotel           0          7              2015               July   
3  Resort Hotel           0         13              2015               July   
4  Resort Hotel           0         14              2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel is_repeated_guest  \
0   BB     PRT         Direct               Direct                 0   
1   BB     PRT         Direct               Direct                 0   
2   BB     GBR         Direct               Direct                 0   
3   BB     GBR      Corporate            Corporate                 0   
4   BB     GBR      Online TA                TA/TO                 0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent company  \
0                  C                3   No Deposit                  
1                  C                4   No Deposit                  
2                  C                0   No Deposit                  
3                  A                0   No Deposit  304.0           
4                  A                0   No Deposit  240.0           

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  \
0                          0          Check-Out              2015-07-01   
1                          0          Check-Out              2015-07-01   
2                          0          Check-Out              2015-07-02   
3                          0          Check-Out              2015-07-02   
4                          1          Check-Out              2015-07-03   

             name                        email  phone-number       credit_card  
0   Ernest Barnes  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1    Andrea Baker       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2  Rebecca Parker   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3    Laura Murray            Laura_M@gmail.com  364-656-8427  ************5677  
4     Linda Hines           LHines@verizon.com  713-226-5883  ************5

In [29]:
data['total_nights'] = data['stays_in_weekend_nights'] + data['stays_in_week_nights']
data.head()

hotel is_canceled  lead_time arrival_date_year arrival_date_month  \
0  Resort Hotel           0        342              2015               July   
1  Resort Hotel           0        737              2015               July   
2  Resort Hotel           0          7              2015               July   
3  Resort Hotel           0         13              2015               July   
4  Resort Hotel           0         14              2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel is_repeated_guest  \
0   BB     PRT         Direct               Direct                 0   
1   BB     PRT         Direct               Direct                 0   
2   BB     GBR         Direct               Direct                 0   
3   BB     GBR      Corporate            Corporate                 0   
4   BB     GBR      Online TA                TA/TO                 0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent company  \
0                  C                3   No Deposit                  
1                  C                4   No Deposit                  
2                  C                0   No Deposit                  
3                  A                0   No Deposit  304.0           
4                  A                0   No Deposit  240.0           

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  \
0                          0          Check-Out              2015-07-01   
1                          0          Check-Out              2015-07-01   
2                          0          Check-Out              2015-07-02   
3                          0          Check-Out              2015-07-02   
4                          1          Check-Out              2015-07-03   

             name                        email  phone-number  \
0   Ernest Barnes  Ernest.Barnes31@outlook.com  669-792-1661   
1    Andrea Baker       Andrea_Baker94@aol.com  858-637-6955   
2  Rebecca Parker   Rebecca_Parker@comcast.net  652-885-2745   
3    Laura Murray            Laura_M@gmail.com  364-656-8427   
4     Linda Hines           LHines@verizon.com  713-226-5883   

        credit_card  total_nights  
0  ************4322             0  
1  ************9157    

In [30]:
stay = data.groupby(['total_nights', 'hotel']).agg('count').reset_index()
stay = stay.iloc[:, :3]
stay = stay.rename(columns={'is_canceled':'No of Stays'})
stay.head()

total_nights         hotel  No of Stays
0             0    City Hotel          308
1             0  Resort Hotel          372
2             1    City Hotel         9169
3             1  Resort Hotel         6580
4             2    City Hotel        10992

In [31]:
fig = px.bar(
    stay,
    x = 'total_nights',
    y = 'No of Stays',
    title = 'Guest stays per Night',
    color = 'hotel',
    barmode = 'group',
    template = 'plotly_dark'
)

fig.show()

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

<h4 align="left"><font color=bdfff6>Insights: Guest stays per night</font></h4>

1. **Most Common Stays**:
   - The majority of stays are **short, ranging from 1 to 4 nights**, for both City and Resort Hotels. This suggests that most guests likely stay for short trips, such as business or weekend leisure trips.
   - For City Hotels:
     - `1 night`: 9,169 stays
     - `2 nights`: 10,992 stays
     - `3 nights`: 11,895 stays
   - For Resort Hotels:
     - `1 night`: 6,580 stays
     - `2 nights`: 4,488 stays
     - `3 nights`: 3,830 stays
   - These shorter stays are the most frequent, especially for `City Hotels`, which may cater to a larger number of business travelers.

2. **Medium-Length Stays (5-7 Nights)**:
   - Stays of **5 to 7 nights** are less common but still notable. `Resort Hotels` have a slightly higher frequency of medium-length stays compared to `City Hotels`, indicating that guests at resorts may be more inclined to stay for a week or longer.
   - `7-night` stays at Resort Hotels see a notable increase (`4,435` stays), possibly indicating weekly vacation packages.

3. **Longer Stays (8+ Nights)**:
   - Stays longer than 8 nights are relatively rare in both types of hotels, with numbers declining sharply as the number of nights increases.
   - Longer stays, especially those exceeding `14 nights`, are mostly seen in `Resort Hotels`, with some exceptional long stays (e.g., 35, 42, 56 nights). This could represent longer vacations or extended stays at resorts.

4. **City vs. Resort Hotel Trends**:
   - `City Hotels` have the highest counts for shorter stays (1–4 nights), aligning with the business traveler profile or short city breaks.
   - `Resort Hotels` have more occurrences of medium and longer stays, which is typical for vacation-oriented locations where guests stay for a full week or even longer.

**Summary**:
Most guests prefer short stays (1–4 nights) at both hotel types, with `City Hotels` seeing a higher concentration of single-night and two-night stays, likely due to business travel. `Resort Hotels` see more medium-length and longer stays, suggesting they attract guests looking for extended vacations.

⬆️ [Tabel of Contents](#0)

## <a id = '3'></a>
<div style = "font-size : 35px; color : #f9858b ; font-family : 'Calibri'; text-align : center; background-color : #bdfff6; border-radius: 5px 5px;"><strong>Ending Note</strong></div> 

### <a id = '3.1'></a>
<div style = "font-size : 25px; color : #ff0099; font-family : 'Comic Sans MS'; "><strong>Exploratory Data Analysis (EDA) Summary</strong></div> 

<div style="font-size: 16px; color: #ff9900; font-family: 'Comic Sans MS';">

Our exploratory data analysis (EDA) provided valuable insights into booking patterns, guest preferences, and stay durations across City and Resort Hotels. Key findings include the most popular months, average daily rates, guest counts, and typical stay durations, which reveal patterns aligned with business travel and leisure stays.

With these insights in mind, the next step is to proceed with `data preprocessing`. This phase will involve:

- `Feature Engineering`: We’ll identify the most relevant features that contribute to meaningful patterns in the dataset, reducing noise and improving model efficiency.
- `Handling Missing Values`: Addressing any missing data to ensure completeness and reliability.
- `Handling Noisy Data`: Cleaning the dataset to reduce inconsistencies or errors.
- `Encoding Categorical Features`: Converting categorical features to numerical values for model compatibility.
- `Normalization/Scaling`: Adjusting numerical features to comparable scales to enhance model accuracy.
- `Handling Imbalanced Data`: Assessing class distributions to determine if resampling techniques are needed.

Completing these preprocessing steps will prepare the data for accurate and robust predictive modeling. 

Let’s proceed with data preprocessing!

⬆️ [Tabel of Contents](#0)